pydantic/transform_experimental.py

```diff
@@ -0,0 +1,506 @@
+"""Experimental functionality, be prepared for this to break."""
+
+from __future__ import annotations
+
+import datetime
+import re
+from collections import deque
+from dataclasses import dataclass
+from decimal import Decimal
+from functools import cached_property
+from typing import TYPE_CHECKING, Annotated, Any, Callable, Generic, Protocol, TypeVar, Union, overload
+
+import annotated_types
+
+if TYPE_CHECKING:
+    from pydantic_core import core_schema as cs
+
+    from pydantic import GetCoreSchemaHandler
+
+from pydantic._internal._internal_dataclass import slots_true
+
+
+@dataclass(**slots_true)
+class _Parse:
+    tp: type[Any]
+    strict: bool = False
+
+
+@dataclass
+class _ParseDefer:
+    func: Callable[[], type[Any]]
+
+    @cached_property
+    def tp(self) -> type[Any]:
+        return self.func()
+
+
+@dataclass(**slots_true)
+class _Transform:
+    func: Callable[[Any], Any]
+
+
+@dataclass(**slots_true)
+class _ValidateOr:
+    left: Validate[Any, Any]
+    right: Validate[Any, Any]
+
+
+@dataclass(**slots_true)
+class _ValidateAnd:
+    left: Validate[Any, Any]
+    right: Validate[Any, Any]
+
+
+_ConstraintAnnotation = Union[
+    annotated_types.Le,
+    annotated_types.Ge,
+    annotated_types.Lt,
+    annotated_types.Gt,
+    annotated_types.Len,
+    annotated_types.MultipleOf,
+    annotated_types.Timezone,
+    annotated_types.Interval,
+    annotated_types.Predicate,
+    re.Pattern[str],
+]
+
+
+@dataclass(**slots_true)
+class _Constraint:
+    constraint: _ConstraintAnnotation
+
+
+_Step = Union[_Parse, _ParseDefer, _Transform, _ValidateOr, _ValidateAnd, _Constraint]
+
+_InT = TypeVar('_InT')
+_OutT = TypeVar('_OutT')
+_NewOutT = TypeVar('_NewOutT')
+
+
+@dataclass(**slots_true)
+class Validate(Generic[_InT, _OutT]):
```

Comment:
Very cool and interesting pattern here. I appreciate the cleanliness of the code and the documentation of major methods, etc - made things quite easy to follow.

I have 2 structural questions:
1. Is there a better way for us to organize the schema application private functions? Not sure if a class based approach would make more sense here, but just something to think about. Ultimately, do we want to build support for this into the `_generate_schema.py` monolith? What would that look like, and how could we smoothly do that?
2. There are a lot of functions that are members of the same category - like str methods, things like that. Right now obviously we've just covered the basic types and transformations, so things aren't super bloated, but I think we should consider potential blowup when we expand to more types in the future. We could use other classes to help categorize these transformations, but that makes accessing such methods more verbose, which goes directly against the intent of this pipeline system.

Reply:
Marking as resolved given the new namespace changes!